### ISSUE TITLE
Add modules to course documents

### ISSUE THREAD

**Issue by draymond63** [uwblueprint/rowan-house#44]
## Summary
The client has expressed the wish to have more functionality in the "folders". For example, these "folders" should have descriptions, and be able to be privatized for staff only. To facilitate this, we are inserting "modules" in-between courses and lessons. Courses will henceforth be the "folder" that contains many modules, and modules will now contain lessons. The current functionality of `courses` will be split between the new `courses` and `modules` appropriately. 

Courses will store modules as a list of objects. Each module will now store a list of references to lessons

## Goals
- [ ] Update `course` documents to follow the new data models in the Notion
- [ ] Add `module` property to lesson doc that stores the ID of the parent module

## Implementation
This code should only be written in the backend. For the data models involved in the database, see the [data models doc](https://www.notion.so/uwblueprintexecs/Data-Models-f6b8178406c1481da654a70e30eabe0e).

## Notes
Some questions from Dan
- How should order be stored? Should we convert this to an array of modules? Should we have an index in each module object specify it's placement?
- Do modules need IDs at all?
- Do lessons need to store their parent module? They are already storing their parent course

**Comment by Shrinjay** [comment]
This entire comment is just me thinking through the design for this, scroll to the end for specific questions and a TLDR. 

Let's think through the data we have and the manipulations we want to do on them:
- Courses have Modules (one to many), Modules have Lessons (one to many, ordered), Lessons have Content (one to many?). Note that we don't know directions for any of these relations.  
- According to the course builder mockups, we can modify Courses, Modules, Lessons and Content.
- We read a module to get its lessons in a specified order, and display the associated content.
- Modifying a course can include altering the properties of the course itself, altering the modules property (id of modules) or the modules themselves. 
- Modifying a module can include altering the properties of the module itself, altering the lessons property (order of modules) or altering the lessons themselves. 
- Modifying a lesson can include altering the properties of the lesson itself or altering the content. 

Therefore, modules need to be stored in a data structure that supports a) one to many relation, with fastest possible reads and updates, from course to module, b) ordered one to many relation, with fastest reads and updates, from module to lessons.

Property a) ad can be satisfied with the current data model, so we should retain the ids as they allow for O(1) operations on independent modules. Property b) requires us to be able to quickly update a specific lesson including its order as well as quickly perform an ordered read. 

The typical ordered data structure in MongoDB is an array, this would support O(1) read (given we know which lesson we want to read), O(1) update of lesson properties, O(1) deletion, but O(n) reordering w/o deletion (with messy code to boot), this is the fastest possible performance for all these operations. However, updating an array is messy in MongoDB and it requires us to store the order of the array as state somewhere in our application (consider what happens if one user changes the order of the array of lessons while another is updating it). Therefore, we want update regardless of order but we want to maintain order on read. 

The best way to satisfy this is to maintain an array that contains lesson IDs in a given order, so when reading, we first read the lesson IDs from the array then fetch the appropriate lessons from the database. In the current data model, modules and lessons are separate MongoDB collections, but I don't think this is really needed if lessons are specific to a given module, in which case we can store the lessons as an embedded document, where the lesson id is the key and the lesson itself is the value. 

**TLDR**
Current Data Model: https://www.notion.so/uwblueprintexecs/Data-Models-f6b8178406c1481da654a70e30eabe0e
_Proposed Data Model_
Course: { ... course properties, modules: { id: Module}} 
Module: {... module properties, ordered_lesson_ids: [array of lesson ids...], lessons: {lesson_id: Lesson}
Lesson: {... lesson properties and content}

Answers to your questions: 
Q1: Maintain order of lessons in an array of lesson ids, same as what we already have.
Q2: Modules need id so we can directly access and modify them in O(1) time. 
Q3: Needs further discussion tbh, the only reason a child would need to care about its parent is if there's some data that in the parent that it depends on, which I don't see happening. Maintaining single direction relations is probably a lot easier. 

Let's discuss this data model specifically, once you approve that I can go ahead and implement it.
